src/pymor/reductors/era.py

```diff
@@ -0,0 +1,160 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.algorithms.projection import project
+from pymor.algorithms.to_matrix import to_matrix
+from pymor.core.cache import cached, CacheableObject
+from pymor.models.iosys import LTIModel
+from pymor.operators.numpy import NumpyHankelOperator, NumpyMatrixOperator
+
+
+class ERAReductor(CacheableObject):
+    r"""Eigensystem Realization Algorithm reductor.
+
+    Constructs a (reduced) realization that matches the Markov parameters optimally for given order
+    by orthogonal factorization of the Hankel matrix of Markov parameters (see :cite:`K78`).
+
+    For a large number if inputs and/or outputs, the factorization of the Hankel matrix can be
+    accelerated by tangential projections of the Markov parameters first (see :cite:`KG16`).
+
+    Attributes
+    ----------
+    data
+        |NumPy array| that contains the first :math:`n` Markov parameters of a discrete-time system.
+        Has to be one- or three-dimensional with either::
+
+            data.shape = (n)
+
+        for scalar-valued Markov parameters or::
+
+            data.shape = (n, p, m)
+
+        for matrix-valued Markov parameters of dimension :math:`p\times m`, where
+        :math:`m` are the number of inputs and :math:`p` are the number of outputs of the system.
+    sampling_time
+        A positive number that denotes the sampling time of the system (in seconds).
+    force_stability
+        Whether the Markov parameters are zero-padded to double the length in order to enforce
+        Kung's stability assumption :cite:`K78`. Defaults to `True`.
+    """
+    cache_region = 'memory'
+
+    def __init__(self, data, sampling_time, force_stability=True):
+        assert sampling_time > 0
```

Comment:
ERA doesn't work for continuous-time systems?

Reply:
Well, usually one uses samples of the impulse response as inputs which are equivalent to the Markov parameters only in discrete-time. 

ERA would work for continuous-time systems, if Markov parameter data is available. I guess with access to a full-order realization one could compute those, but then ERA is equivalent to BT so I did not see the point. It does not hurt to allow it, though.

Reply:
Interesting. If it doesn't hurt, maybe it's ok to allow it. Maybe there is a way to estimate Markov parameters from experimental data?

Reply:
Estimating Markov parameters from frequency domain data is hard I think. In the time domain, it could work since they are the Taylor coefficients for expansion at zero. It still seems very impractical, because you could directly estimate the system with something like `QuadBT` instead of estimating the MPs first.

Generally, I like code to be as general as possible. After thinking again, however, it seems like a big source of error to forget  to set `sampling_time` to a non-zero value. But if you like, I will include it with a logger warning if its zero.

Reply:
Actually, aren't Markov parameters Taylov coefficients of the transfer function at infinity? In that sense, ERA should work the same in continuous and discrete time, right?

Reply:
Yes. I think that it is at impossible/impractical to have access to the Markov parameters in the continuous-time case...

Reply:
I mean, if ERA theoretically works in either case, I would argue that the code should also work fine in both cases (without warnings).

Also, in your work, do you really have measurements of a discrete-time system, or discrete measurements of a continuous-time system? Or does it not matter, since if you measure the impulse response of a continuous-time system $(A, B, C)$ at $t_k = k \Delta t$ for $k = 0, 1, 2, \ldots$, and

$$
C e^{k \Delta t A} B = C \left(e^{\Delta t A}\right)^k B,
$$

means that you obtain Markov parameters of the discrete-time system $(e^{\Delta t A}, B, C)$?

BTW, I don't know if it matters, but [Matlab's `era`](https://www.mathworks.com/help/ident/ref/era.html) seems to allow returning continuous-time systems.

Reply:
> I mean, if ERA theoretically works in either case, I would argue that the code should also work fine in both cases (without warnings).

Ok.

> Or does it not matter, since if you measure the impulse response of a continuous-time system (A,B,C) at tk=kΔt for k=0,1,2,…, and
> 
> CekΔtAB=C(eΔtA)kB,
> 
> means that you obtain Markov parameters of the discrete-time system (eΔtA,B,C)?

Exactly. You usually measure the continuous-time system and conceive of it as discrete-time. Looking at this now, makes me think it might be viable to take the matrix logarithm, at least for diagonalizable A.